pkg/operator/nodecadaemon.go

```diff
@@ -122,65 +121,38 @@ func (c *NodeCADaemonController) sync() error {
 		Status: operatorv1.ConditionUnknown,
 	}
 
-	dsObj, err := gen.Get()
+	_, err := gen.Get()
 	if errors.IsNotFound(err) {
-		availableCondition.Status = operatorv1.ConditionFalse
-		availableCondition.Reason = "NotFound"
-		availableCondition.Message = "The daemon set node-ca does not exist"
+		availableCondition.Status = operatorv1.ConditionTrue
+		availableCondition.Reason = "AsExpected"
+		availableCondition.Message = "The daemon set node-ca is removed"
 
-		progressingCondition.Status = operatorv1.ConditionTrue
-		progressingCondition.Reason = "NotFound"
-		progressingCondition.Message = "The daemon set node-ca does not exist"
+		progressingCondition.Status = operatorv1.ConditionFalse
+		progressingCondition.Reason = "AsExpected"
+		progressingCondition.Message = "The daemon set node-ca is removed"
 	} else if err != nil {
 		availableCondition.Reason = "Unknown"
-		availableCondition.Message = fmt.Sprintf("Unable to check daemon set availability: %s", err)
+		availableCondition.Message = fmt.Sprintf("Unable to check daemon set existence: %s", err)
 
 		progressingCondition.Reason = "Unknown"
-		progressingCondition.Message = fmt.Sprintf("Unable to check daemon set progress: %s", err)
+		progressingCondition.Message = fmt.Sprintf("Unable to check daemon set existence: %s", err)
 	} else {
-		ds := dsObj.(*appsv1.DaemonSet)
-		if ds.Status.NumberAvailable > 0 {
-			availableCondition.Status = operatorv1.ConditionTrue
-			availableCondition.Reason = "AsExpected"
-			availableCondition.Message = "The daemon set node-ca has available replicas"
-		} else {
-			availableCondition.Status = operatorv1.ConditionFalse
-			availableCondition.Reason = "NoAvailableReplicas"
-			availableCondition.Message = "The daemon set node-ca does not have available replicas"
+		gracePeriod := int64(0)
+		propagationPolicy := metav1.DeletePropagationForeground
+		opts := metav1.DeleteOptions{
+			GracePeriodSeconds: &gracePeriod,
+			PropagationPolicy:  &propagationPolicy,
 		}
+		err := gen.Delete(opts)
```

Comment:
What happens if you send the delete request twice?
You can see this:
1. add your own finalizer in the node-ca daemon object
2. run the reconcile loop
3. see what happens when the delete request gets called multiple times

Reply:
After adding a custom finalizer there were no errors in the operator logs. 
After a minute I removed the custom finalizer, and the logs first showed a transition to `NodeCADaemonAvailable: Unable to delete daemon set: daemonsets.apps \"node-ca\" not found`, then immediately transitioned to `NodeCADaemonProgressing: The daemon set node-ca is removed`.
The clusteroperator looks good too:
```
Status:
  Conditions:
    Last Transition Time:  2023-08-11T09:49:22Z
    Message:               Available: The registry is ready
NodeCADaemonAvailable: The daemon set node-ca is removed
ImagePrunerAvailable: Pruner CronJob has been created
    Reason:                Ready
    Status:                True
    Type:                  Available
    Last Transition Time:  2023-08-11T09:49:22Z
    Message:               Progressing: The registry is ready
NodeCADaemonProgressing: The daemon set node-ca is removed
    Reason:                Ready
    Status:                False
    Type:                  Progressing
    Last Transition Time:  2023-08-08T14:50:18Z
    Reason:                AsExpected
    Status:                False
    Type:                  Degraded
```

I wonder if we should change the `NodeCADaemonAvailable` to `NodeCADaemonRemoved` though, but I'm not sure 🤔

Reply:
yep, looks like a race condition - Get returned the object which was deleted before the Delete request

Reply:
I have fixed the race condition. Marking this comment as resolved.